<?xml version="1.0" encoding="utf-8"?>

<robot
  name="ranger_mini_v2" xmlns:xacro="http://ros.org/wiki/xacro">

  <xacro:include filename="$(find gazebo_ranger)/urdf/ranger_l.xacro" />
  <xacro:include filename="$(find gazebo_ranger)/urdf/ranger_r.xacro" />
  <xacro:include filename="$(find gazebo_ranger)/urdf/sensors/kinect_camera.xacro" />

  <!-- Variables -->
  <xacro:property name="M_PI" value="3.14159"/>

  <!-- Vehicle Geometries -->
  <xacro:property name="base_x_size" value="1.0" />
  <xacro:property name="base_y_size" value="0.380000" />
  <xacro:property name="base_z_size" value="0.210000" />

  <xacro:property name="wheelbase" value="0.498" />
  <xacro:property name="track" value="0.58306" />
  <xacro:property name="wheel_vertical_offset" value="-0.0702" />

  <xacro:property name="wheel_length" value="1.1653e-01" />
  <xacro:property name="wheel_radius" value="1.6459e-01" />

  <link name="base_link">
      <inertial>
          <origin xyz="0.0 0.0 0.0"/>
          <mass value="50"/>
          <inertia ixx="1" ixy="0.0" ixz="0.0" iyy="1" iyz="0.0" izz="1"/>
      </inertial>
      <visual>
          <origin xyz="0 0 0" rpy="0 0 0" />
          <geometry>
            <box size="0.5 0.0001 0.2"/>
          </geometry>
          <material
            name="">
            <color
            rgba="0.784 0.784 0.784 1" />
      </material>
      </visual>

  </link>

  <link
    name="ranger_base">
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="file://$(find gazebo_ranger)/meshes/ranger_mini_v2/base_link.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.784 0.784 0.784 1" />
      </material>
    </visual>
  </link>

  <joint
    name="base_to_body_joint"
    type="fixed">
    <origin
      xyz="0.2 0.0 -0.1"
      rpy="0 0 1.57" />
    <parent
      link="base_link" />
    <child
      link="ranger_base" />
    <axis
      xyz="0 0 0" />
  </joint>

  <link
    name="lidar_connector_link">
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="file://$(find gazebo_ranger)/meshes/ranger_mini_v2/lidar_link.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.314 0.314 0.314 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="file://$(find gazebo_ranger)/meshes/ranger_mini_v2/lidar_link.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="lidar_joint"
    type="fixed">
    <origin
      xyz="-0.003125 0.020748 0.40825"
      rpy="1.5708 0 0" />
    <parent
      link="ranger_base" />
    <child
      link="lidar_connector_link" />
    <axis
      xyz="0 0 0" />
  </joint>
  <link
    name="camera_link">
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="file://$(find gazebo_ranger)/meshes/ranger_mini_v2/camera_link.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.314 0.314 0.314 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="file://$(find gazebo_ranger)/meshes/ranger_mini_v2/camera_link.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="camer_joint"
    type="fixed">
    <origin
      xyz="-0.0029998 -0.17935 0.19625"
      rpy="1.5708 0 0" />
    <parent
      link="ranger_base" />
    <child
      link="camera_link" />
    <axis
      xyz="0 0 0" />
  </joint>

  <link
    name="imu_link">
    <visual>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="file://$(find gazebo_ranger)/meshes/ranger_mini_v2/imu_link.STL" />
      </geometry>
      <material
        name="">
        <color
          rgba="0.784 0.784 0.784 1" />
      </material>
    </visual>
    <collision>
      <origin
        xyz="0 0 0"
        rpy="0 0 0" />
      <geometry>
        <mesh
          filename="file://$(find gazebo_ranger)/meshes/ranger_mini_v2/imu_link.STL" />
      </geometry>
    </collision>
  </link>
  <joint
    name="imu_joint"
    type="fixed">
    <origin
      xyz="-0.003 0.087717 0.25475"
      rpy="-1.5708 0 -1.5708" />
    <parent
      link="ranger_base" />
    <child
      link="imu_link" />
    <axis
      xyz="0 0 0" />
  </joint>

   <link name="fl_steering_wheel_link">
        <inertial>
            <origin xyz="0.0 0.0 0.0"/>
            <mass value="1"/>
            <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.01" iyz="0.0" izz="0.01"/>
        </inertial>
        <visual>
            <origin xyz="-0.23 -0.194 0.1" rpy="1.57 0 3.14 " />
            <geometry>
                    <mesh filename="file://$(find gazebo_ranger)/meshes/ranger_mini_v2/fl.dae" />
            </geometry>
        </visual>
        <collision>
            <origin xyz="0 0 0.008" rpy="0 0 0" />
            <geometry>
                <cylinder radius="0.001" length="0.001"/>
            </geometry>
        </collision>
        <collision>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry>
                <cylinder radius="0.001" length="0.001"/>
            </geometry>
        </collision>
    </link>
    <joint name="fl_steering_wheel" type="revolute">
        <origin xyz="0.23 0.206 -0.1" rpy="0 0 0" />
        <parent link="base_link"/>
        <child link="fl_steering_wheel_link"/>
        <axis xyz="0 0 1" rpy="0 0 0" />
        <limit effort="5" velocity="6.28"  lower="-2.1" upper="2.1" />
    </joint>

    <gazebo reference="fl_steering_wheel_link">
      <material>Gazebo/White</material>
    </gazebo>


      <link name="fl_wheel_link">
        <inertial>
            <mass value="8" />
            <origin xyz="0 0 0" />
            <inertia ixx="0.1171" ixy="0" ixz="0" iyy="0.1361" iyz="0" izz="0.1171" />
        </inertial>
        <visual>
            <origin xyz="0 0 0" rpy="1.57 0 0" />
            <geometry>
                <mesh filename="file://$(find gazebo_ranger)/meshes/ranger_mini_v2/ffl_wheel_link.STL" scale="1.15 1.15 1.15" />
            </geometry>
            <material name="fl_wheel">
              <color rgba="1 1 1 1"/>
            </material>
        </visual>
        <collision>
            <origin xyz="0 0 0" rpy="0 1.57 0" />
            <geometry>
                <cylinder radius="0.105" length="0.06"/>
            </geometry>
            
            <contact>
                <ode>
                    <min_depth>0.001</min_depth>
                    <mu1>2000</mu1>
                    <mu2>2000</mu2>
                </ode>
            </contact>
        </collision>
    </link>

    <gazebo reference="fl_wheel_link">
      <material>Gazebo/White</material>
    </gazebo>


    <joint name="fl_wheel" type="continuous">
        <origin xyz="0.017 -0.01 -0.12 " rpy="0 0 1.57" />
        <parent link="fl_steering_wheel_link"/>
        <child link="fl_wheel_link"/>
        <axis xyz="1 0 0" rpy="0 0 0" />
    </joint>


   <link name="fr_steering_wheel_link">
        <inertial>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <mass value="1"/>
            <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.01" iyz="0.0" izz="0.01"/>
        </inertial>
        <visual>
            <origin xyz="-0.23 0.194 0.1" rpy="1.57 0 3.14 " />
            <geometry>
                <mesh filename="file://$(find gazebo_ranger)/meshes/ranger_mini_v2/fr.dae" />
            </geometry>
        </visual>
        <collision>
            <origin xyz="0 0 0.008" rpy="0 0 0" />
            <geometry>
                <cylinder radius="0.001" length="0.001"/>
            </geometry>
        </collision>
    </link>
    <joint name="fr_steering_wheel" type="revolute">
        <origin xyz="0.23 -0.206 -0.1" rpy="0 0 0" />
        <parent link="base_link"/>
        <child link="fr_steering_wheel_link"/>
        <axis xyz="0 0 1" rpy="0 0 0" />
        <limit effort="5" velocity="6.28"  lower="-2.1" upper="2.1" />
    </joint>

    <gazebo reference="fr_steering_wheel_link">
      <material>Gazebo/White</material>
    </gazebo>

    <link name="fr_wheel_link">
        <inertial>
            <mass value="8" />
            <origin xyz="0 0 0" />
            <inertia ixx="0.1171" ixy="0" ixz="0" iyy="0.1361" iyz="0" izz="0.1171" />
        </inertial>
        <visual>
            <origin xyz="0 0 0" rpy="1.57 0 0" />
            <geometry>
                <mesh filename="file://$(find gazebo_ranger)/meshes/ranger_mini_v2/ffl_wheel_link.STL" scale="1.15 1.15 1.15" />
            </geometry>
            <material name="fr_wheel">
              <color rgba="1 1 1 1"/>
            </material>
        </visual>
        <collision>
            <origin xyz="0 0 0" rpy="0 1.57 0" />
            <geometry>
                <cylinder radius="0.105" length="0.06"/>
            </geometry>
                  <contact>
                      <ode>
                    <min_depth>0.001</min_depth>
                    <mu1>2000</mu1>
                    <mu2>2000</mu2>
                      </ode>
                </contact>
        </collision>
    </link>

    <gazebo reference="fr_wheel_link">
      <material>Gazebo/White</material>
    </gazebo>

    <joint name="fr_wheel" type="continuous">
        <origin xyz="0.017 0.01 -0.12 " rpy="0 0 -1.57" />
        <parent link="fr_steering_wheel_link"/>
        <child link="fr_wheel_link"/>
        <axis xyz="-1 0 0" rpy="0 0 0" />
    </joint>


   <link name="rl_steering_wheel_link">
        <inertial>
            <origin xyz="0.0 0.0 0.0" rpy="0.0 0.0 0.0"/>
            <mass value="1"/>
            <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.01" iyz="0.0" izz="0.01"/>
        </inertial>
        <visual>
            <origin xyz="-0.23 -0.194 0.1" rpy="1.57 0 3.14 " />
            <geometry>
                    <mesh filename="file://$(find gazebo_ranger)/meshes/ranger_mini_v2/fl.dae" />
            </geometry>
        </visual>
        <collision>
            <origin xyz="0 0 0.008" rpy="0 0 0" />
            <geometry>
                <cylinder radius="0.001" length="0.001"/>
            </geometry>
        </collision>
    </link>
    <joint name="rl_steering_wheel" type="revolute">
        <origin xyz="-0.23 0.206 -0.1" rpy="0 0 0" />
        <parent link="base_link"/>
        <child link="rl_steering_wheel_link"/>
        <axis xyz="0 0 1" rpy="0 0 0" />
        <limit effort="5" velocity="6.28"  lower="-2.1" upper="2.1" />
    </joint>

    <gazebo reference="rl_steering_wheel_link">
      <material>Gazebo/White</material>
    </gazebo>

        <link name="rl_wheel_link">
            <inertial>
                <mass value="8" />
                <origin xyz="0 0 0" />
                <inertia ixx="0.1171" ixy="0" ixz="0" iyy="0.1361" iyz="0" izz="0.1171" />
            </inertial>
            <visual>
                <origin xyz="0 0 0" rpy="1.57 0 0" />
                <geometry>
                <mesh filename="file://$(find gazebo_ranger)/meshes/ranger_mini_v2/ffl_wheel_link.STL" scale="1.15 1.15 1.15" />
                </geometry>
                <material name="rl_wheel">
                  <color rgba="1 1 1 1"/>
                </material>
            </visual>
            <collision>
            <origin xyz="0 0 0" rpy="0 1.57 0" />
            <geometry>
                <cylinder radius="0.105" length="0.06"/>
            </geometry>
                     <contact>
                         <ode>
                        <mu1>2000</mu1>
                        <mu2>2000</mu2>
                        <min_depth>0.001</min_depth>
                         </ode>
                    </contact>
            </collision>
        </link>

        <joint name="rl_wheel" type="continuous">
          <origin xyz="0.017 -0.01 -0.12 " rpy="0 0 1.57" />
            <parent link="rl_steering_wheel_link"/>
            <child link="rl_wheel_link"/>
            <axis xyz="1 0 0" rpy="0 0 0" />
        </joint>

        <gazebo reference="rl_wheel_link">
          <material>Gazebo/White</material>
        </gazebo>

   <link name="rr_steering_wheel_link">
        <inertial>
            <origin xyz="0.0 0.0 0.0"/>
            <mass value="1"/>
            <inertia ixx="0.01" ixy="0.0" ixz="0.0" iyy="0.01" iyz="0.0" izz="0.01"/>
        </inertial>
        <visual>
            <origin xyz="-0.23 0.194 0.1" rpy="1.57 0 3.14 " />
            <geometry>
                    <mesh filename="file://$(find gazebo_ranger)/meshes/ranger_mini_v2/fr.dae" />
            </geometry>
            
        </visual>
        <collision>
            <origin xyz="0 0 0" rpy="0 0 0" />
            <geometry>
                <cylinder radius="0.001" length="0.001"/>
            </geometry>
        </collision>
    </link>
    <joint name="rr_steering_wheel" type="revolute">
        <origin xyz="-0.23 -0.206 -0.1" rpy="0 0 0" />
        <parent link="base_link"/>
        <child link="rr_steering_wheel_link"/>
        <axis xyz="0 0 1" rpy="0 0 0" />
        <limit effort="5" velocity="6.28"  lower="-2.1" upper="2.1" />
    </joint>

    <gazebo reference="rr_steering_wheel_link">
      <material>Gazebo/White</material>
    </gazebo>

        <link name="rr_wheel_link">
            <inertial>
                <mass value="8" />
                <origin xyz="0 0 0" />
                <inertia ixx="0.1171" ixy="0" ixz="0" iyy="0.1361" iyz="0" izz="0.1171" />
            </inertial>
            <visual>
                <origin xyz="0.0 0.0 0.0" rpy="1.57 0 0" />
                <geometry>
                <mesh filename="file://$(find gazebo_ranger)/meshes/ranger_mini_v2/ffl_wheel_link.STL" scale="1.15 1.15 1.15" />
                </geometry>
                <material name="rr_wheel">
              <color rgba="1 1 1 1"/>
            </material>
            </visual>
            <collision>
            <origin xyz="0 0 0" rpy="0 1.57 0" />
            <geometry>
                <cylinder radius="0.105" length="0.06"/>
            </geometry>
                     <contact>
                         <ode>
                        <min_depth>0.001</min_depth>
                        <mu1>2000</mu1>
                        <mu2>2000</mu2>
                         </ode>
                    </contact>
            </collision>
        </link>

        <gazebo reference="rr_wheel_link">
          <material>Gazebo/White</material>
        </gazebo>

        <joint name="rr_wheel" type="continuous">
            <origin xyz="0.017 0.01 -0.12 " rpy="0 0 -1.57" />
            <parent link="rr_steering_wheel_link"/>
            <child link="rr_wheel_link"/>
            <axis xyz="-1 0 0" rpy="0 0 0" />
        </joint>

	<transmission name="fl_steering_wheel_trans">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="fl_steering_wheel">
		<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      </joint>
	  <actuator name="fl_steering_wheel_motor">
		<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
		<mechanicalReduction>30</mechanicalReduction>
      </actuator>
	</transmission>

	<transmission name="fr_steering_wheel_trans">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="fr_steering_wheel">
		<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      </joint>
	  <actuator name="fr_steering_wheel_motor">
		<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
		<mechanicalReduction>30</mechanicalReduction>
      </actuator>
	</transmission>

	<transmission name="rl_steering_wheel_trans">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="rl_steering_wheel">
		<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      </joint>
	  <actuator name="rl_steering_wheel_motor">
		<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
		<mechanicalReduction>30</mechanicalReduction>
      </actuator>
	</transmission>

	<transmission name="rr_steering_wheel_trans">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="rr_steering_wheel">
		<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
      </joint>
	  <actuator name="rr_steering_wheel_motor">
		<hardwareInterface>hardware_interface/PositionJointInterface</hardwareInterface>
		<mechanicalReduction>30</mechanicalReduction>
      </actuator>
	</transmission>

	<transmission name="fl_wheel_trans">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="fl_wheel">
		<hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
      </joint>
	  <actuator name="fl_wheel_motor">
		<hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
		<mechanicalReduction>30</mechanicalReduction>
      </actuator>
	</transmission>

	<transmission name="fr_wheel_trans">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="fr_wheel">
		<hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
      </joint>
	  <actuator name="fr_wheel_motor">
		<hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
		<mechanicalReduction>30</mechanicalReduction>
      </actuator>
	</transmission>

	<transmission name="rl_wheel_trans">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="rl_wheel">
		<hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
      </joint>
	  <actuator name="rl_wheel_motor">
		<hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
		<mechanicalReduction>30</mechanicalReduction>
      </actuator>
	</transmission>

	<transmission name="rr_wheel_trans">
      <type>transmission_interface/SimpleTransmission</type>
      <joint name="rr_wheel">
		<hardwareInterface>hardware_interface/VelocityJointInterface</hardwareInterface>
      </joint>
	  <actuator name="rr_wheel_motor">
		<hardwareInterface>hardware_interface/PositionJoVelocityJointInterfaceintInterface</hardwareInterface>
		<mechanicalReduction>30</mechanicalReduction>
      </actuator>
	</transmission>

  <gazebo>
    <plugin filename="libgazebo_ros2_control.so" name="gazebo_ros2_control">
      <parameters>$(find gazebo_ranger)/config/config.yaml</parameters>
    </plugin>
  </gazebo>

  <ros2_control name="GazeboSystem" type="system">
      <hardware>
        <plugin>gazebo_ros2_control/GazeboSystem</plugin>
      </hardware>
      
      <joint name="fl_wheel">
        <command_interface name="velocity">
          <param name="min">-3.14</param>
          <param name="max">3.14</param>
        </command_interface>
        <state_interface name="position"/>
        <state_interface name="velocity">
          <param name="initial_value">0.0</param>
        </state_interface>
        <state_interface name="effort"/>
      </joint>

      <joint name="fr_wheel">
        <command_interface name="velocity">
          <param name="min">-3.14</param>
          <param name="max">3.14</param>
        </command_interface>
        <state_interface name="position"/>
        <state_interface name="velocity">
          <param name="initial_value">0.0</param>
        </state_interface>
        <state_interface name="effort"/>
      </joint>

      <joint name="rl_wheel">
        <command_interface name="velocity">
          <param name="min">-3.14</param>
          <param name="max">3.14</param>
        </command_interface>
        <state_interface name="position"/>
        <state_interface name="velocity">
          <param name="initial_value">0.0</param>
        </state_interface>
        <state_interface name="effort"/>
      </joint>

      <joint name="rr_wheel">
        <command_interface name="velocity">
          <param name="min">-3.14</param>
          <param name="max">3.14</param>
        </command_interface>
        <state_interface name="position"/>
        <state_interface name="velocity">
          <param name="initial_value">0.0</param>
        </state_interface>
        <state_interface name="effort"/>
      </joint>

      <joint name="fl_steering_wheel">
        <command_interface name="position">
          <param name="min">-3.14</param>
          <param name="max">3.14</param>
        </command_interface>
        <state_interface name="position">
          <param name="initial_value">0.0</param>
        </state_interface>
        <state_interface name="velocity"/>
        <state_interface name="effort"/>
      </joint>

      <joint name="fr_steering_wheel">
        <command_interface name="position">
          <param name="min">-3.14</param>
          <param name="max">3.14</param>
        </command_interface>
        <state_interface name="position">
          <param name="initial_value">0.0</param>
        </state_interface>
        <state_interface name="velocity"/>
        <state_interface name="effort"/>
      </joint>

      <joint name="rl_steering_wheel">
        <command_interface name="position">
          <param name="min">-3.14</param>
          <param name="max">3.14</param>
        </command_interface>
        <state_interface name="position">
          <param name="initial_value">0.0</param>
        </state_interface>
        <state_interface name="velocity"/>
        <state_interface name="effort"/>
      </joint>

      <joint name="rr_steering_wheel">
        <command_interface name="position">
          <param name="min">-3.14</param>
          <param name="max">3.14</param>
        </command_interface>
        <state_interface name="position">
          <param name="initial_value">0.0</param>
        </state_interface>
        <state_interface name="velocity"/>
        <state_interface name="effort"/>
      </joint>
  </ros2_control>

  <!-- MACRO -->

  <xacro:macro name="box_inertia" params="m x y z">
      <inertia ixx="${m * (y * y + z * z) / 12}" ixy="0" ixz="0" iyy="${m * (x * x + z * z) / 12}" iyz="0" izz="${m * (y * y + x * x) / 12}" />
  </xacro:macro>

  <!-- LIDAR -->
  <xacro:include filename="$(find gazebo_ranger)/urdf/sensors/vlp-32.urdf.xacro"/>
  <xacro:VLP-32 >
    <origin xyz="0 0 0" rpy="-1.57 -1.57 0.0" />
  </xacro:VLP-32> 

  <!-- CAMERA -->
  <xacro:include filename="$(find gazebo_ranger)/urdf/sensors/kinect_camera.xacro" />
  <xacro:kinect_camera camera_key="camera" />

  <!-- IMY -->
  <xacro:include filename="$(find gazebo_ranger)/urdf/sensors/imu.xacro" />
  <xacro:imu />

  <!-- GPS -->
  <xacro:include filename="$(find gazebo_ranger)/urdf/sensors/gps.xacro" />
  <xacro:gps />


</robot>

